File: Medium/2707. Extra Characters in a String/Solution.py
class Solution(object):
    def __init__(self):
        self.dp = []
        self.MX = 100

    def solve(self, s, dictionary, i):
        if i >= len(s):
            return 0

        if self.dp[i] != -1:
            return self.dp[i]

        ans = self.MX

        ans = min(ans, self.solve(s, dictionary, i + 1) + 1)

        for word in dictionary:
            wordSize = len(word)
            if s[i : i + wordSize] == word:
                ans = min(ans, self.solve(s, dictionary, i + wordSize))

        self.dp[i] = ans
        return ans

    def minExtraChar(self, s, dictionary):
        self.dp = [-1] * len(s)
        return self.solve(s, dictionary, 0)

File: Medium/2707. Extra Characters in a String/Solution.cpp
class Solution {
public:

    int dp[51];
    const int MX = 100;

    int solve(string & s, vector<string> &dictionary, int i = 0) {
        if(i >= (int) s.size()) return 0;

        auto & ret = dp[i];
        if(~ret) return ret;

        ret = MX;

        ret = min(ret, solve(s, dictionary, i + 1) + 1);
        for(auto & word : dictionary) {
            int wordSize = (int) word.size();
            string sub = s.substr(i, wordSize);
            if(sub == word)
                ret = min(ret, solve(s, dictionary, i + wordSize));
        }

        return ret;
    }

    int minExtraChar(string s, vector<string>& dictionary) {
        memset(dp, -1, sizeof dp);
        return solve(s, dictionary);
    }
};